controllers/pvc-mutator/encryption/keygen.go

```diff
@@ -0,0 +1,185 @@
+
+	// NamespaceSecretName is the name of the secret containing the user key in
+	// each namespace with encrypted volumes.
+	NamespaceSecretName = "storageos-namespace-key"
+)
+
+var (
+	// ErrCrossNamespace is returned if a encryption key secret is requested
+	// that is not it the PVC namespace.
+	ErrCrossNamespace = errors.New("encryption key secret namespace must match pvc namespace")
+)
+
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+
+// KeyManager is the encrption key manager, responsible for creating and
+// retrieving secrets that contain the keys required for volume encryption.
+type KeyManager interface {
+	Ensure(ctx context.Context, userKeyRef client.ObjectKey, volKeyRef client.ObjectKey) error
+}
+
+// EncryptionKeySetter is responsible for generating and setting pvc encryption
+// keys on a pvc.
+type EncryptionKeySetter struct {
+	// enabledLabel is the pvc label used to indicate that the volume
+	// must have encryption enabled.  It must be set to "true" to enable.
+	enabledLabel string
+
+	// secretNameAnnotationKey is the pvc annotation that stores the name of the
+	// secret containing the encryption key.
+	secretNameAnnotationKey string
+
+	// secretNamespaceAnnotationKey is the pvc annotation that stores the
+	// namespace of the secret containing the encryption key.
+	secretNamespaceAnnotationKey string
+
+	client.Client
+	keys KeyManager
+	log  logr.Logger
+}
+
+// NewEncryptionKeySetter returns a new PVC encryption key mutating admission
+// controller.
+func NewEncryptionKeySetter(k8s client.Client) *EncryptionKeySetter {
+	return &EncryptionKeySetter{
+		enabledLabel:                 EncryptionEnabledLabel,
+		secretNameAnnotationKey:      EncryptionSecretNameAnnotationKey,
+		secretNamespaceAnnotationKey: EncryptionSecretNamespaceAnnotationKey,
+
+		Client: k8s,
+		keys:   keys.New(k8s),
+		log:    ctrl.Log,
+	}
+}
+
+// MutatePVC mutates a given pvc with annotations containing its encryption key,
+// if the pvc has encryption enabled.
+//
+// Errors returned here will block creation of the PVC.
+func (s *EncryptionKeySetter) MutatePVC(ctx context.Context, pvc *corev1.PersistentVolumeClaim, namespace string) error {
+	log := s.log.WithValues("pvc", client.ObjectKeyFromObject(pvc).String())
+	log.V(4).Info("received pvc for mutation")
+
+	// Skip mutation if the PVC does not have encryption enabled.  Don't bother checking
+	// the StorageClass to make sure it's StorageOS.  The encryption label
+	// should only be added to StorageOS PVCs.
+	if !isEnabled(s.enabledLabel, pvc.GetLabels()) {
+		log.V(4).Info("pvc does not have storageos.com/encryption=true label, skipping")
+		return nil
+	}
+
+	// Do not allow secrets from another namespace to be referenced.  We assume
+	// that if the user can create PVCs in the namespace then they should be
+	// able to read volume secrets from that namespace.  We can't assume the
+	// same for other namespaces.
```

Comment:
Yes, that may be needed in the future.  At the moment I'd rather restrict the feature rather than relaxing security constraints.  I think we need to give it a bit more thought and document it properly before allowing it.